src/optimize.ts

```diff
@@ -135,24 +135,28 @@ async function analyse(file: string): Promise<void> {
   //
   let prependToHead = '';
   let appendToHead = '';
+  let heads = $('head'); // always exists because doc loaded with isDocument=true (default)
 
   if (config.html.add_css_reset_as === 'inline') {
     prependToHead += `<style>:where(img){height:auto;}</style>`;
   }
-
   if (prependToHead || appendToHead) {
-    let heads = $('head');
-    if (heads.length === 0) {
-      // head tag doesn't exist
-      $('html')?.prepend(`<head></head>`);
-    }
-    heads = $('head');
-
-    // head tag exist
     if (prependToHead) heads.prepend(prependToHead);
     if (appendToHead) heads.append(appendToHead);
   }
 
+  const charsetElements$ = heads.find('meta[charset]');
+  switch (charsetElements$.length) {
+    case 0: // utf-8 is the default in html5, could check DOCTYPE is set?
+      break;
+    case 1:
+      if (charsetElements$.prev().length) heads.prepend(charsetElements$);
+      break;
+    default:
+      charsetElements$.remove();
```

Comment:
Can you add this?

$state.reportIssue(htmlfile, {
        type: 'fix',
        msg: `Multiple <meta charset> found - taking only the first one and deleting the others.`,
      });

Reply:
(this reporting of issues is garbage but I will refactor it one day)